File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/custom/MoeGate.java
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  *  See the NOTICE file distributed with this work for additional
 *  *  information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

package org.nd4j.linalg.api.ops.impl.transforms.custom;

import lombok.Getter;
import lombok.NoArgsConstructor;
import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.api.ops.DynamicCustomOp;

import java.util.Arrays;
import java.util.List;

/**
 * Mixture-of-Experts (MoE) gating operation.
 * <p>
 * Computes top-K expert routing with load-balancing auxiliary loss:
 * <pre>
 *   logits         = hiddenStates @ gateWeight
 *   expertIndices  = topK(softmax(logits))
 *   gateWeights    = normalized weights for selected experts
 *   auxLoss        = load balancing loss
 * </pre>
 * <p>
 * Inputs:
 * <ul>
 *   <li>0: hiddenStates [T, H] - token hidden states</li>
 *   <li>1: gateWeight [H, E] - gating projection (H=hidden, E=numExperts)</li>
 * </ul>
 * <p>
 * Integer arguments:
 * <ul>
 *   <li>0: topK (default 2) - number of experts per token</li>
 *   <li>1: numExperts (default 8) - total number of experts</li>
 * </ul>
 * <p>
 * Float arguments:
 * <ul>
 *   <li>0: auxLossCoeff (default 0.01) - load balancing loss coefficient</li>
 * </ul>
 * <p>
 * Outputs:
 * <ul>
 *   <li>0: expertIndices [T, K] INT64 - selected expert indices</li>
 *   <li>1: gateWeights [T, K] - routing weights for selected experts</li>
 *   <li>2: auxLoss [1] - auxiliary load balancing loss scalar</li>
 * </ul>
 *
 * @author Eclipse Deeplearning4j Contributors
 */
@NoArgsConstructor
public class MoeGate extends DynamicCustomOp {

    @Getter private int topK = 2;
    @Getter private int numExperts = 8;
    @Getter private float auxLossCoeff = 0.01f;

    /**
     * SameDiff constructor with default options.
     */
    public MoeGate(SameDiff sameDiff, SDVariable hiddenStates, SDVariable gateWeight) {
        super(null, sameDiff, new SDVariable[]{hiddenStates, gateWeight}, false);
        addIArgument((long) topK, (long) numExperts);
        addTArgument((double) auxLossCoeff);
    }

    /**
     * SameDiff constructor with double auxLossCoeff.
     */
    public MoeGate(SameDiff sameDiff, SDVariable hiddenStates, SDVariable gateWeight,
                   int topK, int numExperts, double auxLossCoeff) {
        this(sameDiff, hiddenStates, gateWeight, topK, numExperts, (float) auxLossCoeff);
    }

    /**
     * Full SameDiff constructor with all options.
     */
    public MoeGate(SameDiff sameDiff, SDVariable hiddenStates, SDVariable gateWeight,
                   int topK, int numExperts, float auxLossCoeff) {
        super(null, sameDiff, new SDVariable[]{hiddenStates, gateWeight}, false);
        this.topK = topK;
        this.numExperts = numExperts;
        this.auxLossCoeff = auxLossCoeff;
        addIArgument((long) topK, (long) numExperts);
        addTArgument((double) auxLossCoeff);
    }

    /**
     * INDArray constructor.
     */
    public MoeGate(INDArray hiddenStates, INDArray gateWeight,
                   INDArray expertIndices, INDArray gateWeights, INDArray auxLoss,
                   int topK, int numExperts, float auxLossCoeff) {
        super(null, new INDArray[]{hiddenStates, gateWeight},
                expertIndices != null ? new INDArray[]{expertIndices, gateWeights, auxLoss} : null);
        this.topK = topK;
        this.numExperts = numExperts;
        this.auxLossCoeff = auxLossCoeff;
        addIArgument((long) topK, (long) numExperts);
        addTArgument((double) auxLossCoeff);
    }

    public MoeGate(SameDiff sd, SDVariable input, SDVariable gateWeights, int numExperts, int topK) {
        super(null, sd, new SDVariable[]{input, gateWeights}, false);
        this.topK = topK;
        this.numExperts = numExperts;
        addIArgument((long) topK, (long) numExperts);
        addTArgument((double) this.auxLossCoeff);
    }

    public MoeGate(INDArray input, INDArray gateWeights, int numExperts, int topK) {
        super(new INDArray[]{input, gateWeights}, null);
        this.topK = topK;
        this.numExperts = numExperts;
        addIArgument((long) topK, (long) numExperts);
        addTArgument((double) this.auxLossCoeff);
    }

    @Override
    public void configureFromArguments() {
        super.configureFromArguments();
        if (iArguments.size() > 0) this.topK = iArguments.get(0).intValue();
        if (iArguments.size() > 1) this.numExperts = iArguments.get(1).intValue();
        if (tArguments.size() > 0) this.auxLossCoeff = tArguments.get(0).floatValue();
    }

    @Override
    public String opName() {
        return "moe_gate";
    }

    @Override
    public List<DataType> calculateOutputDataTypes(List<DataType> inputDataTypes) {
        DataType dt = inputDataTypes.get(0);
        return Arrays.asList(DataType.INT64, dt, dt);
    }

    @Override
    public int getNumOutputs() {
        return 3;
    }
}

File: libnd4j/include/ops/declarable/generic/nn/moe_gate.cpp
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  * See the NOTICE file distributed with this work for additional
 *  * information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

//
// @author Eclipse Deeplearning4j
//

#include <system/op_boilerplate.h>

#if NOT_EXCLUDED(OP_moe_gate)

#include <helpers/ConstantShapeHelper.h>
#include <helpers/MmulHelper.h>
#include <ops/BroadcastOpsTuple.h>
#include <ops/declarable/headers/nn.h>
#include <ops/declarable/helpers/activations.h>
#include <ops/declarable/helpers/one_hot.h>
#include <ops/declarable/helpers/top_k.h>

namespace sd {
namespace ops {

CUSTOM_OP_IMPL(moe_gate, 2, 3, false, -2, -2) {
    auto hiddenStates = INPUT_VARIABLE(0);  // [tokens, hidden]
    auto gateWeight   = INPUT_VARIABLE(1);  // [hidden, numExperts]

    auto expertIndices = OUTPUT_VARIABLE(0);  // [tokens, topK] INT64
    auto gateWeights   = OUTPUT_VARIABLE(1);  // [tokens, topK]
    auto auxLoss       = OUTPUT_VARIABLE(2);  // [1]

    const LongType topK       = block.getIArguments()->size() > 0 ? INT_ARG(0) : 2;
    const LongType numExperts = block.getIArguments()->size() > 1 ? INT_ARG(1) : gateWeight->sizeAt(1);
    const double auxLossCoeff = block.getTArguments()->size() > 0 ? T_ARG(0) : 0.01;

    REQUIRE_TRUE(hiddenStates->rankOf() == 2, 0,
                 "moe_gate: hiddenStates must be rank 2 [tokens, hidden], got rank %i", hiddenStates->rankOf());
    REQUIRE_TRUE(gateWeight->rankOf() == 2, 0,
                 "moe_gate: gateWeight must be rank 2 [hidden, numExperts], got rank %i", gateWeight->rankOf());
    REQUIRE_TRUE(gateWeight->sizeAt(0) == hiddenStates->sizeAt(1), 0,
                 "moe_gate: gateWeight rows (%lld) must match hiddenStates hidden dim (%lld)",
                 (long long)gateWeight->sizeAt(0), (long long)hiddenStates->sizeAt(1));
    REQUIRE_TRUE(gateWeight->sizeAt(1) == numExperts, 0,
                 "moe_gate: gateWeight columns (%lld) must equal numExperts (%lld)",
                 (long long)gateWeight->sizeAt(1), (long long)numExperts);
    REQUIRE_TRUE(topK > 0 && topK <= numExperts, 0,
                 "moe_gate: topK (%lld) must be in [1, numExperts=%lld]", (long long)topK, (long long)numExperts);

    if (hiddenStates->isEmpty()) {
        auxLoss->p(0, 0.0);
        return Status::OK;
    }

    const LongType numTokens = hiddenStates->sizeAt(0);

    // router logits [tokens, numExperts], then softmax over the expert axis.
    // softmax needs a distinct output buffer (stateful transform).
    std::vector<sd::LongType> logitsShape = {numTokens, numExperts};
    NDArray logits('c', logitsShape, hiddenStates->dataType(), block.launchContext());
    NDArray probs('c', logitsShape, hiddenStates->dataType(), block.launchContext());
    MmulHelper::mmul(hiddenStates, gateWeight, &logits, 1.0, 0.0);
    helpers::softmax(block.launchContext(), &logits, &probs, 1);

    // top-K expert selection per token
    std::vector<sd::LongType> selShape = {numTokens, topK};
    NDArray topVals('c', selShape, hiddenStates->dataType(), block.launchContext());
    auto tkStatus = helpers::topKFunctor(block.launchContext(), &probs, &topVals, expertIndices, topK, true);
    REQUIRE_TRUE(tkStatus == Status::OK, 0, "moe_gate: top-K selection failed");

    // renormalize the selected probabilities so each token's weights sum to 1
    std::vector<LongType> lastDim = {1};
    NDArray* rowSum = topVals.reduceAlongDimension(reduce::Sum, &lastDim, true);  // [tokens, 1]
    topVals.applyTrueBroadcast(BroadcastOpsTuple::Divide(), rowSum, gateWeights, true);

    // load-balancing auxiliary loss: coeff * E * sum_e(meanProb_e * routedFrac_e)
    std::vector<LongType> axis0 = {0};
    NDArray* meanProbs = probs.reduceAlongDimension(reduce::Mean, &axis0, false);  // [numExperts]

    std::vector<sd::LongType> oneHotShape = {numTokens, topK, numExperts};
    NDArray oneHot('c', oneHotShape, hiddenStates->dataType(), block.launchContext());
    helpers::onehot(block.launchContext(), expertIndices, &oneHot, 2, numExperts, 1.0, 0.0);
    std::vector<LongType> axes01 = {0, 1};
    NDArray* routedFrac = oneHot.reduceAlongDimension(reduce::Mean, &axes01, false);  // [numExperts]

    *meanProbs *= *routedFrac;
    NDArray* balance = meanProbs->reduceNumber(reduce::Sum);
    auxLoss->p(0, auxLossCoeff * static_cast<double>(numExperts) * balance->e<double>(0));

    delete rowSum;
    delete meanProbs;
    delete routedFrac;
    delete balance;

    return Status::OK;
}

DECLARE_TYPES(moe_gate) {
    getOpDescriptor()
        ->setAllowedInputTypes(0, {ALL_FLOATS})
        ->setAllowedInputTypes(1, {ALL_FLOATS})
        ->setAllowedOutputTypes(0, {ALL_INTS})
        ->setAllowedOutputTypes(1, {ALL_FLOATS})
        ->setAllowedOutputTypes(2, {ALL_FLOATS})
        ->addTraits(OP_TRAIT_FULLY_WRITING);
}

DECLARE_SHAPE_FN(moe_gate) {
    auto hsShape = inputShape->at(0);
    auto inputType = sd::ArrayOptions::dataType(hsShape);

    const LongType numTokens = shape::sizeAt(hsShape, static_cast<sd::LongType>(0));
    const LongType topK = block.getIArguments()->size() > 0 ? INT_ARG(0) : 2;

    std::vector<sd::LongType> selShape = {numTokens, topK};
    auto indicesShapeInfo = ConstantShapeHelper::getInstance().createShapeInfo(sd::DataType::INT64, 'c', selShape);
    auto weightsShapeInfo = ConstantShapeHelper::getInstance().createShapeInfo(inputType, 'c', selShape);

    std::vector<sd::LongType> lossShape = {1};
    auto lossShapeInfo = ConstantShapeHelper::getInstance().createShapeInfo(inputType, 'c', lossShape);

    return SHAPELIST(indicesShapeInfo, weightsShapeInfo, lossShapeInfo);
}

}  // namespace ops
}  // namespace sd

#endif
